### PR TITLE
executor: track the memory usage of IndexLookUpExecutor

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

File: executor/distsql.go [hunk 1]
```diff
@@ -521,6 +539,13 @@ func (e *IndexLookUpExecutor) Open(ctx context.Context) error {
 }
 
 func (e *IndexLookUpExecutor) open(ctx context.Context, kvRanges []kv.KeyRange) error {
+	// We have to initialize "memTracker" and other execution resources in here
+	// instead of in function "Open", because this "IndexLookUpExecutor" may be
+	// constructed by a "IndexLookUpJoin" and "Open" will not be called in that
+	// situation.
+	e.memTracker = memory.NewTracker(e.id, e.ctx.GetSessionVars().MemQuotaSort)
```
Comment (coocood) [inline executor/distsql.go]: Why use MemQuotaSort?

Comment (zz-jason) [reply executor/distsql.go]: add another session variable named `tidb_mem_quota_indexlookupreader` to control this.

File: executor/distsql.go [hunk 2]
```diff
@@ -831,50 +872,65 @@ func (w *tableWorker) pickAndExecTask(ctx context.Context) {
...
 	tableReader, err = w.buildTblReader(ctx, task.handles)
 	if err != nil {
 		log.Error(err)
-		return
+		return errors.Trace(err)
 	}
 	defer terror.Call(tableReader.Close)
+
+	task.memTracker = w.memTracker
+	memUsage := int64(cap(task.handles))
```
Comment (coocood) [inline executor/distsql.go]: A handle takes 8 bytes.

File: executor/distsql.go [hunk 3]
```diff
@@ -831,50 +872,65 @@ func (w *tableWorker) pickAndExecTask(ctx context.Context) {
...
 		}
+		memUsage = chk.MemoryUsage()
+		task.memUsage += memUsage
+		task.memTracker.Consume(memUsage)
 		iter := chunk.NewIterator4Chunk(chk)
 		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
 			task.rows = append(task.rows, row)
 		}
 	}
+	memUsage = int64(cap(task.rows))
```
Comment (coocood) [inline executor/distsql.go]: A chunk Row takes 16 bytes.

File: executor/distsql.go [hunk 4]
```diff
@@ -66,6 +67,20 @@ type lookupTableTask struct {
...
+	// memTracker is used to release memUsage after task is done and unused.
+	//
+	// The sequence of function calls are:
+	//   1. calculate task.memUsage.
+	//   2. task.memTracker = tableWorker.memTracker
+	//   3. task.memTracker.Consume(task.memUsage)
+	//   4. task.memTracker.Consume(-task.memUsage)
+	//
+	// Step 1~2 are completed in "tableWorker.executeTask".
+	// Step 3~4 are completed in "IndexLookUpExecutor.NextChunk".
```
Comment (coocood) [inline executor/distsql.go]: Step 3 is completed in executeTask.

File: executor/distsql.go [hunk 5]
```diff
@@ -691,9 +720,15 @@ func (e *IndexLookUpExecutor) getResultTask() (*lookupTableTask, error) {
 	if !ok {
 		return nil, nil
 	}
-	err := <-task.doneCh
-	if err != nil {
-		return nil, errors.Trace(err)
+	for err := range task.doneCh {
```
Comment (coocood) [inline executor/distsql.go]: Why close the `doneCh` channel and use for loop to receive it?

Comment (zz-jason) [reply executor/distsql.go]: refactored.

File: executor/distsql.go [hunk 6]
```diff
@@ -896,7 +886,7 @@ func (w *tableWorker) executeTask(ctx context.Context, task *lookupTableTask) er
 	defer terror.Call(tableReader.Close)
 
 	task.memTracker = w.memTracker
-	memUsage := int64(cap(task.handles))
+	memUsage := int64(cap(task.handles) + 8)
```
Comment (coocood) [inline executor/distsql.go]: * 8

**Comment by shenli** [comment]
Please fix the CI.

**Comment by zz-jason** [comment]
@coocood @XuHuaiyu PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**